clients/src/main/java/org/apache/kafka/clients/RecordHeaders.java

```diff
@@ -0,0 +1,177 @@
+            this.headers = new ArrayList<>();
+        } else if (headers instanceof RecordHeaders) {
+            //new ArrayList using efficient Arrays.copyOf the backing list in given RecordHeaders
+            this.headers = new ArrayList<>(((RecordHeaders) headers).headers);
+        } else if (headers instanceof Collection) {
+            //new ArrayList using efficient Arrays.copyOf the given collection.
+            this.headers = new ArrayList<>((Collection<Header>) headers);
+        } else {
+            //new ArrayList iterating through to copy the headers.
+            this.headers = new ArrayList<>();
+            Iterator<Header> iterator = headers.iterator();
+            while (iterator.hasNext()) {
+                this.headers.add(iterator.next());
+            }
+        }
+    }
+
+    @Override
+    public Headers add(Header header) throws IllegalStateException {
+        if (isClosed) {
+            throw new IllegalStateException("RecordHeaders isClosed");
+        }
+        headers.add(header);
+        return this;
+    }
+
+    @Override
+    public Headers remove(String key) throws IllegalStateException {
+        if (isClosed) {
+            throw new IllegalStateException("RecordHeaders isClosed");
+        }
+        Iterator<Header> iterator = iterator();
+        while (iterator.hasNext()) {
+            if (iterator.next().key().equals(key)) {
+                iterator.remove();
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public Header lastHeader(String key) {
+        for (int i = headers.size() - 1; i >= 0; i--) {
+            Header header = headers.get(i);
+            if (header.key().equals(key)) {
+                return header;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Iterable<Header> headers(final String key) {
+        return new Iterable<Header>() {
+            @Override
+            public Iterator<Header> iterator() {
+                return RecordHeaders.this.filter(headers.iterator(), key);
+            }
+        };
+    }
+
+    @Override
+    public Iterator<Header> iterator() {
+        return closeAware(headers.iterator());
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.isClosed = true;
+    }
+
+    private Iterator<Header> closeAware(final Iterator<Header> original) {
+        return new Iterator<Header>() {
+            @Override
+            public boolean hasNext() {
+                return original.hasNext();
+            }
+
+            public Header next() {
+                return original.next();
+            }
+
+            @Override
+            public void remove() {
+                if (isClosed) {
+                    throw new IllegalStateException("RecordHeaders isClosed");
+                }
+                original.remove();
+            }
+        };
+    }
+
+    private Iterator<Header> filter(final Iterator<Header> original, final String key) {
```

Comment:
Yes this one could have been, it was using java 8 feature previous, was simply quickly removing our usage of java 8, as kip 118 isn't in master and thus would cause a build failure as still needing java 7 support atm. 

I will change to static inner class for now.